src/main/java/org/terasology/web/resources/worldMap/WorldMapResource.java

```diff
@@ -0,0 +1,232 @@
+    @Override
+    protected ResourceMethod<WorldMapInput, String> getPutMethod(ResourcePath path) throws ResourceAccessException {
+        return createParameterlessMethod(path, ClientSecurityRequirements.PUBLIC, WorldMapInput.class,
+                (data, client) -> getWorldMapBase64ImageString(data.getCenter(), data.getMapBlockWidth(), data.getMapBlockLength(), data.isSurface()));
+    }
+
+    /**
+     * Get the color of blocks in the world map and convert them into a base64 encoded image.
+     * @param center the location of the block in the center of the map.
+     * @param mapBlockWidth the width of the map.
+     * @param mapBlockLength the length of the map.
+     * @param isSurface whether or not to get blocks on the surface (using the block at the highest elevation that isn't air).
+     * @return a base64 encoded png image of the world map with location and size depending on the parameters.
+     */
+    private String getWorldMapBase64ImageString(Vector3i center, int mapBlockWidth, int mapBlockLength, boolean isSurface) {
+        final int colorSizeMultiplier = mapBlockWidth * mapBlockLength <= 125 * 125 ? 60 : 30;
+        int blockY = BLOCK_Y_DEFAULT;
+        List<List<Color>> colors = new ArrayList<>(mapBlockWidth);
+        for (int i = 0; i < mapBlockWidth; ++i) {
+            colors.add(i, new ArrayList<>(mapBlockLength));
+        }
+
+        loadChunks(center, mapBlockWidth, mapBlockLength);
+
+        for (int x = (int) Math.floor((double) center.getX() - mapBlockWidth / 2); x < (int) Math.ceil((double) mapBlockWidth / 2 + center.getX()); ++x) {
+            for (int z = (int) Math.floor((double) center.getZ() - mapBlockLength / 2); z < (int) Math.ceil((double) mapBlockLength / 2 + center.getZ()); ++z) {
+                while (worldProvider.getBlock(x, blockY, z).getURI().toString().equals("engine:unloaded")) {
+                    try {
+                        Thread.sleep(1000);
+                    } catch (InterruptedException e) {
+                        e.printStackTrace();
+                    }
+                }
+                blockY = isSurface ? getSurfaceY(x, blockY, z) : center.getY();
+                Block block = worldProvider.getBlock(x, blockY, z);
+                ResourceUrn blockUrn = block.getURI().getBlockFamilyDefinitionUrn();
+                if (Assets.get(blockUrn, BlockTile.class).isPresent()) {
+                    BufferedImage blockImage = Assets.get(blockUrn, BlockTile.class).get().getImage();
+                    colors.get(x - (center.getX() - mapBlockWidth / 2)).add(new Color(getColorOfTexture(blockImage, blockY)));
+                } else {
+                    logger.warn("cannot find texture of block " + blockUrn.toString());
+                    colors.get(x - (center.getX() - mapBlockWidth / 2)).add(Color.BLACK);
+                }
+            }
+        }
+
+        int colorSize = (int) (1 / Math.floor(Math.log(mapBlockWidth * mapBlockLength)) * colorSizeMultiplier);
```

Comment:
Two things:
1) It took me a while to understand this is the length of the side of a square representing a block in the image, I had to get to line 117 to get it - and in my understanding it hasn't much to do with colors, maybe this could be renamed to something more meaningful like `blockSize` or `blockSide`?
2) I don't fully understand the formula, personally I think I'h have just used a constant size (like `colorSizeMultiplier`). If I understand correctly `mapBlockWidth * mapBlockLength` is the area measured in blocks, but I don't understand why you compute it's logarithm and take it's inverse, could you please explain what's the reason?

Reply:
I think renaming the variable to `imageBlockSize` would work.

The formula is used because otherwise, the image displayed on the frontend would be too large for a 250x250 map, and too small for a 20x20 map. The formula works to normalize the size of the images. It's the best way I could find to do it, but if you have any suggestions for a better solution then I would be glad to hear it.

Reply:
Ah okay thanks for the explanation. It works fine and I can't think of any better way right now, so I agree for keeping your formula